fireworks/core/mongo_launchpad.py

```diff
@@ -0,0 +1,1407 @@
+        self.workflows.delete_many({})
+        #self.offline_runs.delete_many({})
+        self._restart_ids(1, 1)
+        if self.gridfs_fallback is not None:
+            self.db.drop_collection("{}.chunks".format(GRIDFS_FALLBACK_COLLECTION))
+            self.db.drop_collection("{}.files".format(GRIDFS_FALLBACK_COLLECTION))
+        self.tuneup()
+        self.m_logger.info('LaunchPad was RESET.')
+
+    def _get_duplicates(self, fw_id, include_self=False, allowed_states=None):
+        if type(allowed_states) == str:
+            allowed_states = {'$in': allowed_states}
+        f = self.fireworks.find_one({"fw_id": fw_id, "spec._dupefinder": {"$exists": True}},
+                                    {'launches':1})
+        query = {"launches": {"$in": f['launches']}}
+        if allowed_states:
+            query['state'] = allowed_states
+        if not include_self:
+            query['fw_id'] = {"$ne": fw_id}
+        if f:
+            for d in self.fireworks.find(query, {"fw_id": 1}):
+                duplicates.append(d['fw_id'])
+        return list(set(duplicates))
+
+    def _recover(self, fw_id, launch_idx=None):
+        """
+        Function to get recovery data for a given fw
+        Args:
+            fw_id (int): fw id to get recovery data for
+            launch_id (int or 'last'): launch_id to get recovery data for, if 'last'
+                recovery data is generated from last launch
+        """
+        if launch_idx is not None:
+            m_fw = self.get_fw_by_id(fw_id, launch_idx)
+            recovery = m_fw.state_history[-1].get("checkpoint")
+            recovery.update({'_prev_dir': m_fw.launch_dir,
+                             '_launch_idx': m_fw.launch_idx})
+            # Launch recovery
+            recovery.update({'_mode': recover_mode})
+            set_spec = {'$set': {'spec._recovery': recovery}}
+            if recover_mode == 'prev_dir':
+                prev_dir = m_fw.launch_dir
+                set_spec['$set']['spec._launch_dir'] = prev_dir
+            self.fireworks.find_one_and_update({"fw_id": fw_id}, set_spec)
+
+        # If no launch recovery specified, unset the firework recovery spec
+        else:
+            set_spec = {"$unset":{"spec._recovery":""}}
+            self.fireworks.find_one_and_update({"fw_id":fw_id}, set_spec)
+
+    def _get_launch_by_id(self, launch_id):
+        """
+        Given a Launch id, return launch info.
+
+        Args:
+            launch_id (int): launch id
+
+        Returns:
+            dict
+        """
+        m_launch = self.launches.find_one({'launch_id': launch_id})
+        if m_launch:
+            m_launch["action"] = get_action_from_gridfs(m_launch.get("action"), self.gridfs_fallback)
+            if 'launch_id' in m_launch:
+                m_launch.pop('launch_id')
+            return m_launch
+        raise ValueError('No Launch exists with launch_id: {}'.format(launch_id))
+
+    def _get_launch_by_fw_id(self, fw_id, launch_idx=None):
```

Comment:
Does this get used anywhere?

Reply:
Yes, by get_fw_dict_by_id and _delete_wf